### PR TITLE
Allow to postpone message delivery to `Aggregate` and `ProcessManager` instances

### PR DESCRIPTION
**Pull request by armiol**
Within this PR a more flexible way to control the message delivery to `Aggregate`s and `ProcessManager`s is introduced.

Previously it was only possible to set a custom delivery strategy on a `Bus` level. It helped to achieve a distributed delivery of messages. 

For instance in case write-side and read-side were deployed to different nodes in cloud — using the `Bus`-level delivery one could do the trick:
* postpone `Event`s making their ways to `ProjectionRepository` instances on the write-side, 
* store their envelopes to a cloud persistent storage,
* and manually route them to the instance, responsible for the read-side.

Another real-world case was to balance the `Command` payload targeting `AggregateRepositories`. All the commands targeting the same instance of `Aggregate` had to be grouped and executed on the same instance — to avoid concurrency issues ruining the event log.

When `@React` came along, the routing possibilities became much more complex. Also it became possible to modify the `Aggregate` state by listening to events, as well as commands previously. The `Bus`-level delivery no longer was flexible enough — it didn't have a notion of target entity ID. Hence it was nearly impossible to postpone and group messages to avoid concurrency issues.

This PR implements an ability to postpone delivery on per-`Aggregate`- or  per-`ProcessManager`-per-ID basis. 

It became possible to group all the messages, leading to the entity state modification of  `Aggregate` (or `ProcessManager`) instance, postpone their chaotic delivery and apply all of them to a single entity instance later — when it makes sense and with no concurrency issues.

To do that it is required to implement custom delivery strategy by extending one of the `@SPI` abstract classes:

* `AggregateEventDelivery` — for customizing the event delivery to `Aggregate` instances.
* `AggregateRejectionDelivery` — for customizing the rejection delivery to `Aggregate` instances.
* `AggregateCommandDelivery` — for customizing the command delivery to `Aggregate` instances.
* `PmEventDelivery` — for customizing the event delivery to `ProcessManager` instances.
* `PmRejectionDelivery` — for customizing the rejection delivery to `ProcessManager` instances.
* `PmCommandDelivery` — for customizing the command delivery to `ProcessManager` instances.

`AggregateRepository` and `ProcessManagerRepository` have a number of methods, returning the default non-postponing strategy implementations for events, rejections and commands. So to customize this behavior, an appropriate method should be overridden:

```
// Custom `Event` delivery strategy for `Aggregate`s.
public class CustomEventDelivery
        extends AggregateEventDelivery<ProjectId, ProjectAggregate> {

    protected CustomEventDelivery(ProjectRepository repository) {
        super(repository);
    }

    @Override
    public boolean shouldPostpone(ProjectId id, EventEnvelope envelope) {
       // define a strategy here, basing upon aggregate ID and the envelope.
    }
}

// Repository definition with an overridden strategy for `Event` delivery.
public class ProjectRepository extends AggregateRepository<ProjectId, ProjectAggregate> {

    private CustomEventDelivery delivery = ...;

    @Override
    public PostponingEventDelivery getEventEndpointDelivery() {
        return delivery;      
    }
}
```

To deliver the postponed events a `deliverNow(entityId, envelope)` method should be used:
```

final CustomEventDelivery delivery = ...;
//...
delivery.deliverNow(storedEventData.getProjectId(), storedEventData.getEventEnvelope());
```

### CODE REVIEW

**Comment by armiol** [comment]
@alexander-yevsyukov PTAL. I am not going to merge anything anyway, so that's not urgent.

File: server/src/main/java/io/spine/server/aggregate/AggregateEventEndpoint.java [hunk 1]
```diff
@@ -52,6 +57,12 @@ private AggregateEventEndpoint(AggregateRepository<I, A> repo, EventEnvelope eve
     }
 
     @Override
+    protected AggregateEndpointDelivery<I, A, EventEnvelope> getEndpointDelivery(
```
Comment (alexander-yevsyukov) [inline server/src/main/java/io/spine/server/aggregate/AggregateEventEndpoint.java]: Please fix wrapping.

File: server/src/main/java/io/spine/server/procman/PmEndpointDelivery.java [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+package io.spine.server.procman;
+
+import io.spine.core.ActorMessageEnvelope;
+import io.spine.server.delivery.EndpointDelivery;
+
+/**
+ * @author Alex Tymchenko
```
Comment (alexander-yevsyukov) [inline server/src/main/java/io/spine/server/procman/PmEndpointDelivery.java]: Please add Javadoc.

File: server/src/main/java/io/spine/server/procman/PmCommandDelivery.java [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+ */
+package io.spine.server.procman;
+
+import io.spine.annotation.SPI;
+import io.spine.core.CommandEnvelope;
+
+/**
+ * A strategy on delivering the ecommandsvents to the instances of a certain process manager type.
+ *
+ * @author Alex Tymchenko
```
Comment (alexander-yevsyukov) [inline server/src/main/java/io/spine/server/procman/PmCommandDelivery.java]: Please document generic parameters.

File: server/src/main/java/io/spine/server/delivery/EndpointDelivery.java [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+ * A strategy on delivering the messages to the instances of a certain aggregate type.
+ *
+ * <p>Allows {@linkplain #shouldPostpone(Object, ActorMessageEnvelope) to postpone} the delivery
+ * at runtime for a certain ID and message.
+ *
+ * <p>The postponed messages are not dispatched to the entity instances automatically. However
+ * it is expected they are dispatched manually later via
+ * {@linkplain #deliverNow(Object, ActorMessageEnvelope) deliverNow(ID, envelope)} method call.
+ *
+ * @author Alex Tymchenko
```
Comment (alexander-yevsyukov) [inline server/src/main/java/io/spine/server/delivery/EndpointDelivery.java]: Please document generic parameters.

File: server/src/main/java/io/spine/server/procman/PmEventDelivery.java [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+ */
+package io.spine.server.procman;
+
+import io.spine.annotation.SPI;
+import io.spine.core.EventEnvelope;
+
+/**
+ * A strategy on delivering the events to the instances of a certain process manager type.
+ *
+ * @author Alex Tymchenko
```
Comment (alexander-yevsyukov) [inline server/src/main/java/io/spine/server/procman/PmEventDelivery.java]: Please document generic parameters.

File: server/src/main/java/io/spine/server/procman/PmRejectionDelivery.java [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+ */
+package io.spine.server.procman;
+
+import io.spine.annotation.SPI;
+import io.spine.core.RejectionEnvelope;
+
+/**
+ * A strategy on delivering the rejections to the instances of a certain process manager type.
+ *
+ * @author Alex Tymchenko
```
Comment (alexander-yevsyukov) [inline server/src/main/java/io/spine/server/procman/PmRejectionDelivery.java]: Please document generic parameters.

File: server/src/main/java/io/spine/server/procman/PmEndpointDelivery.java [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+ */
+public abstract class PmEndpointDelivery<I,
```
Comment (alexander-yevsyukov) [inline server/src/main/java/io/spine/server/procman/PmEndpointDelivery.java]: `@SPI`?

**Comment by armiol** [comment]
@alexander-yevsyukov PTAL. Issues are addressed.

**Pull request merged**